### ISSUE TITLE
False positive on indentation for Annotation Arrays

### ISSUE THREAD

**Issue by s-ste** [checkstyle/checkstyle#17122]
I have read the documentation: https://checkstyle.org/checks/misc/indentation.html#Indentation
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac -cp src/main/java src/main/java/com/example/checkstyledemo/Bar.java
... compiles successfully

/var/tmp $ cat config/checkstyle/checkstyle.yml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">

    <property name="charset" value="UTF-8"/>

    <property name="severity" value="${org.checkstyle.google.severity}" default="warning"/>

    <property name="fileExtensions" value="java"/>
    <module name="TreeWalker">
        <module name="Indentation">
            <property name="basicOffset" value="2"/>
            <property name="braceAdjustment" value="2"/>
            <property name="caseIndent" value="2"/>
            <property name="throwsIndent" value="4"/>
            <property name="lineWrappingIndentation" value="4"/>
            <property name="arrayInitIndent" value="2"/>
        </module>
    </module>
</module>

/var/tmp $ cat src/main/java/com/example/checkstyledemo/Bar.java
package com.example.checkstyledemo;

@AnnotList({
    @Annot(
      "hello"
    ),
    @Annot(
      "world"
    )
})
public class Bar {
}


/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.24.0-all.jar -c config/checkstyle/checkstyle.xml src/main/java/com/example/checkstyledemo/Bar.java
Starting audit...
[WARN] Bar.java:9:5: 'annotation array initialization' child has incorrect indentation level 4, expected level should be one of the following: 6, 8. [Indentation]
[WARN] Bar.java:12:5: 'annotation array initialization' child has incorrect indentation level 4, expected level should be one of the following: 6, 8. [Indentation]
Audit done.
```

---

I have also created a reproducer repository based on Spring linked here https://github.com/s-ste/checkstyle-indentation-reproducer

---

Hi everybody, I recently upgraded Checkstyle from 10.23.1 to 10.24.0, and now I get warnings that the linked code doesn't follow the rules anymore. I've compared my code to the release notes and found some parts related to fixes in indentation, but I don't see a direct connection so far.

I would expect Checkstyle to continue accepting the above code or make it configurable to allow this. Maybe I'm missing something, but to me it doesn't make much sense to indent the closing parenthesis here.

google checks config of indentation is same:
https://github.com/checkstyle/checkstyle/blob/60e9f192a73fc0f071150212eaaa3f2d503f476d/src/main/resources/google_checks.xml#L300-L307

**Comment by romani** [comment]
@mohitsatr, is this something we can quickly fix to unblock users ?

**Comment by s-ste** [comment]
I had a look in the history and found that the behaviour has changed with commit edbfc66ebec525c024ea967c82a72c38fe2b7fa2 which indeed reworked indentation.

**Comment by mohitsatr** [comment]
> is this something we can quickly fix to unblock users ?

on it.

**Comment by mohitsatr** [comment]
not sure If a quick fix will work here. It's always tricky with Indentation

**Comment by mohitsatr** [comment]
@romani @Zopsss can you help make sense of the following:

```
@AnnotList({ indent=0
    @Annot( // indent=4 , but arrayInitIndent is 2
      "hello" // indent=6, is it lineWrappingIndentation ?
    ), // indent=4
```
why does second line have indentation of 4, clearly it isn't the case of `lineWrappingIndentation` and even `lineWrappingIndentation` would not have been valid here because it is case of `arrayInitIndent`. 
we also need to update to docs to include examples for `arrayInitIndent` property

**Comment by mohitsatr** [comment]
ping @s-ste https://github.com/checkstyle/checkstyle/pull/17135#issuecomment-2906804457

**Comment by romani** [comment]
We need to understand intention of author of issue.
If he wanted to follow Google style, than following of google-java-format is required.

If config of Indentation Check is custom, than we remove all labels of relationships to Google style, and continue discussion without google-java-format.

@s-ste , please confirm what is your config.

**Comment by romani** [comment]
it is array:
```
$ cat Test.java 
package com.example.checkstyledemo;

@AnnotList({
    @Annot(
      "hello"
    ),
    @Annot(
      "world"
    )
})
public class Bar {
}

$ java -jar checkstyle-10.23.0-all.jar -t Test.java
COMPILATION_UNIT -> COMPILATION_UNIT [1:0]
|--PACKAGE_DEF -> package [1:0]
|   |--ANNOTATIONS -> ANNOTATIONS [1:19]
|   |--DOT -> . [1:19]
|   |   |--DOT -> . [1:11]
|   |   |   |--IDENT -> com [1:8]
|   |   |   `--IDENT -> example [1:12]
|   |   `--IDENT -> checkstyledemo [1:20]
|   `--SEMI -> ; [1:34]
`--CLASS_DEF -> CLASS_DEF [3:0]
    |--MODIFIERS -> MODIFIERS [3:0]
    |   |--ANNOTATION -> ANNOTATION [3:0]
    |   |   |--AT -> @ [3:0]
    |   |   |--IDENT -> AnnotList [3:1]
    |   |   |--LPAREN -> ( [3:10]
    |   |   |--ANNOTATION_ARRAY_INIT -> { [3:11]
    |   |   |   |--ANNOTATION -> ANNOTATION [4:4]
    |   |   |   |   |--AT -> @ [4:4]
    |   |   |   |   |--IDENT -> Annot [4:5]
    |   |   |   |   |--LPAREN -> ( [4:10]
    |   |   |   |   |--EXPR -> EXPR [5:6]
    |   |   |   |   |   `--STRING_LITERAL -> "hello" [5:6]
    |   |   |   |   `--RPAREN -> ) [6:4]
    |   |   |   |--COMMA -> , [6:5]
    |   |   |   |--ANNOTATION -> ANNOTATION [7:4]
    |   |   |   |   |--AT -> @ [7:4]
    |   |   |   |   |--IDENT -> Annot [7:5]
    |   |   |   |   |--LPAREN -> ( [7:10]
    |   |   |   |   |--EXPR -> EXPR [8:6]
    |   |   |   |   |   `--STRING_LITERAL -> "world" [8:6]
    |   |   |   |   `--RPAREN -> ) [9:4]
    |   |   |   `--RCURLY -> } [10:0]
    |   |   `--RPAREN -> ) [10:1]
    |   `--LITERAL_PUBLIC -> public [11:0]
    |--LITERAL_CLASS -> class [11:7]
    |--IDENT -> Bar [11:13]
    `--OBJBLOCK -> OBJBLOCK [11:17]
        |--LCURLY -> { [11:17]
        `--RCURLY -> } [12:0]

```

**Comment by Zopsss** [comment]
Did you confirmed the code using google-java-formatter? If yes then please show it here, then we can treat this issue as "google-style" related.

**Comment by Zopsss** [comment]
> just so that I understand, does that mean that only behavior that deviates from the `google-java-format` tool is buggy?

@s-ste, as @mohitsatr said at https://github.com/checkstyle/checkstyle/issues/17122#issuecomment-2924192405

> Google-style does not specify the handling of annotation arrays.

We're just formatting the code using google-java-formatter tool to format your provided code and see how the newly formatted code looks and does it gives violation with our google config or not, if it gives violation then we need to check which tool is wrong here, Checkstyle or google-java-formatter, if Checkstyle is wrong then this issue will be treated as "google-style" related issue.

To confirm the styling, we can open issue on [google-java-formatter](https://github.com/google/google-java-format)'s repo

**Comment by mohitsatr** [comment]
https://github.com/checkstyle/checkstyle/actions/runs/15224263896/job/42824293794?pr=17135

**Comment by romani** [comment]
@mohitsatr , formater just made all single line as it is possible.  
user provided most minimalistic code to show a problem.

```
/var/tmp$ cat Test.java 
package com.example.checkstyledemo;

@AnnotList({
    @Annot(
      "hello, sdhgksj hgkj hkjgdh kgjshkjghdkjdhjkdhgkjhdjhkjghdhfgjkshfjghdjfhgkjdfhjghdfjgdfh"
    ),
    @Annot(
      "hello2, sdhgksj hgkj hkjgdh kgjshkjghdkjdhjkdhgkjhdjhkjghdhfgjkshfjghdjfhgkjdfhjghdfjgdfh"
    )
})
public class Bar {
}

$ java -jar google-java-format-1.27.0-all-deps.jar Test.java > TestFormatted.java
$ diff -u Test.java TestFormatted.java 
--- Test.java	2025-05-31 12:15:39.860844760 -0700
+++ TestFormatted.java	2025-05-31 12:18:27.698325343 -0700
@@ -1,12 +1,9 @@
 package com.example.checkstyledemo;
 
 @AnnotList({
-    @Annot(
-      "hello, sdhgksj hgkj hkjgdh kgjshkjghdkjdhjkdhgkjhdjhkjghdhfgjkshfjghdjfhgkjdfhjghdfjgdfh"
-    ),
-    @Annot(
-      "hello2, sdhgksj hgkj hkjgdh kgjshkjghdkjdhjkdhgkjhdjhkjghdhfgjkshfjghdjfhgkjdfhjghdfjgdfh"
-    )
+  @Annot(
+      "hello, sdhgksj hgkj hkjgdh kgjshkjghdkjdhjkdhgkjhdjhkjghdhfgjkshfjghdjfhgkjdfhjghdfjgdfh"),
+  @Annot(
+      "hello2, sdhgksj hgkj hkjgdh kgjshkjghdkjdhjkdhgkjhdjhkjghdhfgjkshfjghdjfhgkjdfhjghdfjgdfh")
 })
-public class Bar {
-}
+public class Bar {}

$ java -jar checkstyle-10.24.0-all.jar -c /google_checks.xml TestFormatted.java
Starting audit...
[WARN] /var/tmp/TestFormatted.java:3:1: The name of the outer type and the file do not match. [OuterTypeFilename]
[WARN] /var/tmp/TestFormatted.java:3:1: Missing a Javadoc comment. [MissingJavadocType]
Audit done.

$ java -jar checkstyle-10.24.0-all.jar -c /google_checks.xml Test.java
Starting audit...
[WARN] /var/tmp/Test.java:3:1: The name of the outer type and the file do not match. [OuterTypeFilename]
[WARN] /var/tmp/Test.java:3:1: Missing a Javadoc comment. [MissingJavadocType]
[WARN] /var/tmp/Test.java:6:5: 'annotation array initialization' child has incorrect indentation level 4, expected level should be one of the following: 6, 8. [Indentation]
[WARN] /var/tmp/Test.java:9:5: 'annotation array initialization' child has incorrect indentation level 4, expected level should be one of the following: 6, 8. [Indentation]
Audit done.

```

So no violations after formatting.
but indentation is changed.

user provided code is not Google formatted compliant, BUT validation by Indentation is still confusing as suggested indentation is not correct.

So it is valid google style issue,  There should be violations, but messages of violations should be improved.

**Comment by s-ste** [comment]
Hi, I wasn't aware of the google-java-formatter tool. In that case I can see that my code isn't compliant with the Google code style. I'll have another look if I can achieve my intended result with checkstyle and will let you know.

Thanks for looking into it so far :)

**Comment by romani** [comment]
@mohitsatr , please create new issue as summary of successor of this issue. in new issue Please make relationship to google formatter behavior and problem should be wrong confusing messages.

**Comment by tsologub** [comment]
I re-read this thread five times and still can't connect the dots. Please help. I am using the Google `checkstyle.xml` the latest version and the Checkstyle the latest version. I am receiving the same issues as the topic starter. Though when switching to Checkstyle 10.23 everything works great.

**Comment by mohitsatr** [comment]
@tsologub It's always a good idea to create an issue with the code and config you're using, so we can take a proper look and help fix it.

**Comment by romani** [comment]
Fix is merged

**Comment by romani** [comment]
@s-ste , @tsologub , 10.26.0 is released , please test on your side, lets us know if something else is still not working well.
